### PR TITLE
Import shadowing (under review)

### PR DESCRIPTION
**Pull request by gergoerdi**
Local binders are allowed to shadow names defined in outer scopes. Here, we propose that top-level binders should also be allowed to shadow imported names, basically treating imported modules as living in a scope that is "outside" the top-level one.

[Rendered here](https://github.com/gergoerdi/ghc-proposals/blob/mu/import-shadowing/proposals/0000-import-shadowing.rst)

### CODE REVIEW

File: proposals/0000-import-shadowing.rst [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+Local binders are allowed to shadow names defined in outer
+scopes. Here, we propose that top-level binders should also be allowed
+to shadow imported names, basically treating imported modules as
+living in a scope that is "outside" the top-level one.
+
+
+Motivation
+----------
+Currently, there is no way in Haskell to shadow an imported name. The
```
Comment (goldfirere) [inline proposals/0000-import-shadowing.rst:24]: This sentence tripped me. Because you _can_ shadow an imported name, at least locally. Maybe it would be better to have the word "globally" in this sentence?

File: proposals/0000-import-shadowing.rst [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+in the definition of ``goFishing`` resolves to the definition in
+``Example1`` even when ``catch`` is not explicitly hidden from the
+import list of ``Control.Exception``.
+
+We believe this is an improvement to the status quo because it is a
+natural extension of the idea behind the shadowing policy of local
+binders: that shadowing expresses the intention by the user to only
+care about the names that are defined "more nearby".
+
+Proposed Change Specification
```
Comment (goldfirere) [inline proposals/0000-import-shadowing.rst:67]: This section does not read as a _specification_ to me. Instead, I read it as an interpretation of an unwritten specification by means of desugaring, and that desugaring is itself specified only by example.

Happily, I think the specification is very easy:

* Occurrences of an identifier in the code of a module and in its export list are looked up in an environment containing all the global definitions in that module. If that lookup fails, then the identifier is looked up in an environment containing all the definitions imported by `import` statements (including the implicit `import` of the `Prelude`, if any).

I think that covers it!

File: proposals/0000-import-shadowing.rst [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+ module Mod where
+
+ zip = ...
+
+The above example is taken directly from `the "Import" page of the
+Haskell Wiki <https://wiki.haskell.org/Import>`_.
+   
+Effect and Interactions
+-----------------------
+Beside intra-module references, the other place where top-level
```
Comment (goldfirere) [inline proposals/0000-import-shadowing.rst:219]: This looks like an extra bit of specification, not just an effect or interaction. I included this in my little specification bullet, above.

**Comment by gergoerdi** [comment]
> It's probably worth mentioning explicitly how this interacts with `-Wname-shadowing`.

I think since this *is* shadowing, it should trigger the appropriate warning as well, i.e. if `-Wname-shadowing` is turned on, it should warn about top-level bindings that shadow imported names.

**Review by goldfirere — COMMENTED** [review]
I'm in strong support of this. Coding now some in OCaml, I'm really enjoying its different approach to shadowing.

I'm curious, though: does Mu support shadowing from one `import` over another? That is, in OCaml, if I have

```ocaml
open A
open B
```

then anything `B` exports shadows anything `A` exports. Contrast to Haskell, where both imports go into the same environment, potentially causing ambiguity.

I have come to prefer the shadowing semantics, because it's more expressive. In particular, it means that, if there are some definitions I positively do _not_ want (e.g. `head`), I can shadow them with definitions that are sure not to type-check. (This is quite common in the OCaml I work with!) It also means that users importing several libraries with the same name do not need to carefully hide or quality exports, etc. Instead, they have to carefully order their imports. That's still work, yes, but I think it's less work to just order the imports than to make sure to list every identifier that they wish to be unambiguous.

So I guess I'm arguing not only in favor of this proposal as written, but actually to strengthen it to allow later imports to shadow earlier ones.

To be clear, I'd be in support of the proposal even without that strengthening.

**Comment by gergoerdi** [comment]
> I'm curious, though: does Mu support shadowing from one `import` over another?

Wouldn't that be a much more expansive change? There are code formatters for sorting imports, some people like grouping imports "semantically", etc., all this depends on imports being permutation-invariant.

**Comment by tomjaguarpaw** [comment]
> > I'm curious, though: does Mu support shadowing from one import over another?
>
> Wouldn't that be a much more expansive change? There are code formatters for sorting imports, some people like grouping imports "semantically", etc., all this depends on imports being permutation-invariant.

Yeah, Haskell's permutation invariance of definitions is an extremely nice property and I can't see myself using anything that violates that. See also my objection on https://github.com/ghc-proposals/ghc-proposals/pull/401#issuecomment-781521535.

**Comment by gergoerdi** [comment]
> I can't see myself using anything that violates that

I don't want to let this slide -- I do think ordered imports can be good, I enjoy Agda's ability to even interleave imports and definitions (which of course makes a lot more sense if your toplevel scope is, unlike in Haskell, a `let` not a `letrec`). 

*But* I think it's not a good idea *to add this to Haskell*.

File: proposals/0000-import-shadowing.rst (hunk 1 above)
Comment (yuri-sq) [reply proposals/0000-import-shadowing.rst:24]: This is completely independent feature
```haskell
-- the original is about (pseudoHaskell)
let import A (a)
    import B (b)
in  a = shadowing

-- yours is about
let import A (a)
in  let import B (a)
    in  a = shadowing
```

Comment (gergoerdi) [reply proposals/0000-import-shadowing.rst:24]: > This is completely independent feature

I agree, supposing you meant this to be a reply not to this sub-thread, but @goldfirere's original top-level comment.

**Comment by alt-romes** [comment]
> A second-order effect of this is that using this extension can lead to preemptive forward compatibility. Adding a new export to Prelude can lead to breakage just by virtue of existing code defining and using top-level definitions with the same name. With ImportShadowing, the existing intra-module references keep their meaning and there is no migration needed to accomodate the new Prelude names.

This is quite a nice property in favour of this proposal.

However, if it's a library shadowing a Prelude name and exporting it from a given module, users of said library will still have ambiguity when importing both the library and Prelude. For definitions in that module, `ImportsShadowing` does forward-proof it against changes in Prelude.

**Comment by adamgundry** [comment]
Given the recent addition of `foldl'` to `Prelude`, the justification for this is pretty compelling.

I agree that making imports order-dependent is more ambiguous. Another option would be to have explicit import lists shadow imports without an import list, for example if we have `import M (foo)` and `import N` (where `N` exports `foo`), then references to `foo` could be resolved to `M.foo`. And perhaps the implicit `Prelude` import is "even more implicit" than `import M`...

**Comment by VitWW** [comment]
Your proposal breaks current usage of import.
But we could have an alternative: special pseudo-keyword on import

```haskell
import shadows Control.Exception;
-- import Control.Exception hiding (catch)

catch :: Fish -> IO Food
catch fish = ...
```

File: proposals/0000-import-shadowing.rst (hunk 1 above)
Comment (goldfirere) [reply proposals/0000-import-shadowing.rst:24]: :+1: to "shadow an imported name in the top-level scope"

**Comment by goldfirere** [comment]
To be clear, I'm happy to withdraw my [little extra proposal](https://github.com/ghc-proposals/ghc-proposals/pull/652#pullrequestreview-2057798186) about ordered imports, though I shall hold it in wait for another, more opportune day.

I continue to support the original proposal.

**Comment by Ericson2314** [comment]
Hehe, and my first impression (years ago) was one of horror at how order-dependent OCaml felt :D

**Comment by Ericson2314** [comment]
I think it would be good to specify how reexports work.

```haskell
module Foo ( module Bar ) where

import Bar

-- bar shadows
bar = ... 
```

I think I would want this to be an error, rather than export the shadowed `bar` or the local `bar` without an error.

**Comment by simonpj** [comment]
> I think it would be good to specify how reexports work.

The [Haskell report](https://www.haskell.org/onlinereport/haskell2010/haskellch5.html#x11-1000005.2) says "The form “module M” names the set of all entities that are in scope with both an unqualified name “e” and a qualified name “M.e”."

There is no need to change that.  In the example you give, the export item `module Bar` would export the `bar` from `Bar`.

But well worth pointing this out in the proposal.

A couple of other points

* Please add section numbers; it's a one-line change in the header.
* The "Proposed change specification" seems more complicated, and less precise, than necessary. How about this:
  * Consider an occurrence of an unqualified name `x`, not bound locally (by let, lambda, case etc):
     * If there is a top-level binding of `x` then the occurrence is resolved to that binding, regardless of `import` declarations.
     * Otherwise, if the `import` declarations bring into scope a unique entity with unqualified name `x`, the occurrence is resolved to that entity.
  * Consider an occurrence of a qualified name `M.x`:
    * If the module is called `M` and there is a top-level binding of `x`, the occurrence is resolved to that binding.
     * Otherwise, if the `import` declarations bring into scope a unique entity with unqualified name `M.x`, the occurrence is resolved to that entity.

**Comment by AntC2** [comment]
**Transitive Imports vs direct Imports**

I didn't get a clear sense for this scenario (apologies if it is covered).

* Your module imports `A` (which includes `fun`);
* then defines its own `fun`.
* Within your module, `fun` denotes `Your.fun` no ambiguity.

* My module imports `Your`;
* doesn't explicitly import `A`; but it gets imported transitively.
* Presumably unqualified `fun` means `Your.fun`?

* I change my module to explicitly import `A` as well as `Your`.
* But no other changes.
* Now unqualified `fun` is ambiguous?

**Comment by phadej** [comment]
@AntC2 do you mean, that at the end

```haskell
module A (fun) where

fun = 'a'
```

```haskell
module Your (fun) where

import A

fun = 'y
```

```haskell
module My where

import A
import Your

expr = ... fun ...
```

?

As far as I understand the proposal the use of `fun` in `My.expr` is ambiguous. That's what @goldfirere hijack-proposes to disambiguate based on `import` order. In an example above `fun` in `My` would be resolved to `Your.fun`. People in the discussion don't like it, as currently we are free to reorder the imports.

In my opinion, in this case GHC should *warn early* that `A` and `Your` import clashing names, *before* they are actually used and ambiguity cannot be resolved. I.e.

```haskell
module My2 where
import A
import Your
```

would compile with a warning like

```
warning: imports A and Your bring ambiguous `fun` into the scope. 
```

but this warning is orthogonal to this proposal.

---

In summary, as far as I understand, the proposal helps with forward-compatibility when the target module shadows the definitions from imports (e.g. `Data.Map` defining `foldl'`). It doesn't help with the case when there are multiple open imports (e.g. implicit `Prelude` and `import Data.Map`) and there are no shadowing (module uses a `foldl'`).

**Comment by gergoerdi** [comment]
> Your proposal breaks current usage of import.

I'm proposing to gate this under a new language extension `ImportShadowing`, not just changing the meaning of `import` for all modules.

**Comment by gergoerdi** [comment]
> * Your module imports `A` (which includes `fun`);
> * then defines its own `fun`.
> * Within your module, `fun` denotes `Your.fun` no ambiguity.
> * My module imports `Your`;
> * doesn't explicitly import `A`; but it gets imported transitively.

If you just import `Your`, you don't transitively import `A` just because `Your` imports it. Only names exported by `Your` are brought into scope, so no conflict on `fun`.

> * I change my module to explicitly import `A` as well as `Your`.
> * But no other changes.
> * Now unqualified `fun` is ambiguous?

Yes, this is correct. This proposal is about top-level bindings in the current module vs. imported names. Your scenario is about imported names vs imported names.

**Comment by gergoerdi** [comment]
> * Please add section numbers; it's a one-line change in the header.

I'm not sure what you mean here. Is this some RestructuredText feature I need to somehow turn on? I'm currently using the standard Proposal template.

I like your reformulation of the spec, but this is tripping me up:

>     * Otherwise, if the `import` declarations bring into scope a unique entity with unqualified name `M.x`, the occurrence is resolved to that entity.

There can be things with the *unqualified* name `M.x`?!

**Comment by simonpj** [comment]
> I'm not sure what you mean here. Is this some RestructuredText feature I need to somehow turn on? I'm currently using the standard Proposal template.

Aargh.  Yes, you can only do this with `.rst` not `.md`.   It is really really worth it  -- could you possibly convert?

@adamgundry I think we should get rid of the `.md` template.  It just causes trouble.

**Comment by simonpj** [comment]
> Otherwise, if the `import` declarations bring into scope a unique entity with unqualified name `M.x`, the occurrence is resolved to that entity.

Oh sorry, that was a copy/paste-o.  It should say "qualified name M.x" of course.

**Comment by gergoerdi** [comment]
> Aargh. Yes, you can only do this with `.rst` not `.md`. It is really really worth it -- could you possibly convert?

But I am ~Pagliacci~ already using the `.rst` template.

**Comment by AntC2** [comment]
Thanks @phadej, @gergoerdi, I had more in mind where these were in a sense the same `fun`:

```
module Your (fun) where

import A

fun | ... exotic condition = ... special handling
    | otherwise            = A.fun
```

> would compile with a warning like
>
>        warning: imports A and Your bring ambiguous `fun` into the scope. 
>
> but this warning is orthogonal to this proposal.
>

Yeah, OK this is current behaviour and the proposal isn't making it worse. `Your` module should warn importers to hide `A.fun`.

**Comment by gergoerdi** [comment]
@simonpj 

> The "Proposed change specification" seems more complicated, and less precise, than necessary. How about this:

Thanks, I've rewritten it based on this, let me know if it makes sense. I'm not sure about the "can be resolved" language but I wanted to write something that makes sense in the Haskell 2010 setting as well:

> Consider an occurrence of an unqualified name ``x``, not bound locally
> (by `let`, lambda, a `case` alternative, etc). There are two
> possible sources of resolving it:
> 
> (A) If there is a top-level binding of `x` then the occurrence can be
>     resolved to that binding.
> 
> (B) If the import declarations bring into scope a unique entity with
>     unqualified name `x`, the occurrence can be resolved to that entity.
> 
> Consider an occurrence of a qualified name `M.x`:
> 
> (A) If the module is called `M` and there is a top-level binding of
>     ``x``, the occurrence can be resolved to that binding
> 
> (B) If the import declarations bring into scope a unique entity with
>     qualified name `M.x`, the occurrence can be resolved to that
>     entity.
> 
> In both cases, Haskell 2010 regards cases (A) and (B) on equal
> footing: if exactly one of the two cases can be used to resolve the
> name, that case is used; if both cases can be used, then the
> occurrence is ambiguous and reported as such.
> 
> This proposal instead tries (A) and (B) in order, i.e. if the (A) case
> resolves the occurrence, then that is used, and the (B) case is only
> checked otherwise.

**Comment by simonpj** [comment]
> But I am Pagliacci already using the .rst template.

Aaargh. So you are.  I have no idea what is happening.  Can anyone help?  Section numbering has disappeared from **all proposals** despite the 
```
.. sectnum::
```
incantation. @nomeata knows about this kind of thing.

**Comment by simonpj** [comment]
> Thanks, I've rewritten it based on this, let me know if it makes sense.

Much better. I'd say "is resolved" rather than "can be resolved".

**Comment by gergoerdi** [comment]
@simonpj :

> Section numbering has disappeared from **all proposals** 

in that case, I think this is merely a GitHub rendering issue.

**Comment by gergoerdi** [comment]
@simonpj :

> Much better. I'd say "is resolved" rather than "can be resolved".

Should I leave in or move somewhere else or remove the `let`-desugaring explanation? As a reader, I know I would find it very helpful in explaining the issue at a glance.

**Comment by gergoerdi** [comment]
@simonpj the CI rendering at https://ghc-proposals--652.org.readthedocs.build/en/652/proposals/0000-import-shadowing.html does show section numbers, fyi.

**Comment by simonpj** [comment]
> Should I leave in or move somewhere else or remove the let-desugaring explanation? As a reader, I know I would find it very helpful in explaining the issue at a glance.

If you find it helpful, leave it in.  But rather than describe it as the *specification* I'd put it in a subsection called "An alternative way to understand the specification".  That way, a reader knows that they *can* read the section, but they don't *have* to read and understand it -- it's optional.  Very helpful to distinguish normative material from optional illustration.

File: proposals/0000-import-shadowing.rst [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+So this is a "-1"-impact change: it doesn't break existing code, and
+"un-breaks" existing broken code.
+
+A second-order effect of this is that using this extension can lead to
+preemptive forward compatibility. Adding a new export to ``Prelude``
+can lead to breakage just by virtue of existing code defining and
+using top-level definitions with the same name. With
+``ImportShadowing``, the existing intra-module references keep their
+meaning and there is no migration needed to accomodate the new
+``Prelude`` names.
```
Comment (noughtmare) [inline proposals/0000-import-shadowing.rst:320]: Consider moving this to the motivation section

File: proposals/0000-import-shadowing.rst [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+
+Alternatives
+------------
```
Comment (noughtmare) [inline proposals/0000-import-shadowing.rst:323]: You could also mention @goldfirere later imports shadow earlier imports as an alternative.

File: proposals/0000-import-shadowing.rst (hunk 4 above)
Comment (gergoerdi) [reply proposals/0000-import-shadowing.rst:320]: Does it belong in *Motivation* even if it wasn't the motivating idea, more of a gimme?

Comment (noughtmare) [reply proposals/0000-import-shadowing.rst:320]: Whether or not it was part of your original motivation to write this proposal doesn't really matter in my opinion. I think the intention of the motivation section should be to convince the committee to accept the proposal and to explain to everyone who looks back on this in the future why we made this choice.

Currently, I think the motivation is kind of meager in that your only argument seems to be that it "is a natural extension", i.e. it makes the language more consistent. That feels a bit abstract to me, I'd like to see more concrete reasons like the extra forward compatibility and, now that I think about it, the stackoverflow questions in the endorsements section (which also seem out of place there, because those people are not endorsing this proposal specifically).

**Comment by phadej** [comment]
@adamgundry 

> Given the recent addition of foldl' to Prelude, the justification for this is pretty compelling.

But this proposal doesn't help with that as much as it could. As the new behavior is guarded behind enabling an extension, downstream users will need to enable this extension to get its benefit (either pro or retroactively). I doubt that many would do the change proactively (I probably wouldn't)

An alternative is to have the proposed behavior as default, and just warn about non-Haskell2010 like shadowing. As proposal already states, that won't change the meaning of valid Haskell2010 programs.

**Comment by gergoerdi** [comment]
> An alternative is to have the proposed behavior as default, and just warn about non-Haskell2010 like shadowing (in `-Wdefault`). As proposal already states, that won't change the meaning of valid Haskell2010 programs.

Should this be part of the proposal, or should it be just one more line item in some general "which extensions should be on by default" discussion?

**Comment by JakobBruenker** [comment]
@maralorn note that `-Wname-shadowing` is not enabled by default, so (with the proposal as is), the compiler will not warn by default.

**Comment by maralorn** [comment]
Yeah, that's fine. This warning is about a situation which will very rarely lead to a subtle runtime bug without Type error. Thus the warning can (remain) opt in.

**Comment by Tritlo** [comment]
I think this is good, and the "matches expected behavior on Stack Overflow" is a great argument for adoption. I think you should explicitly mention the interaction with the name shadowing warning, I was surprised to not see it mentioned.

**Comment by gergoerdi** [comment]
@Tritlo 

> I think you should explicitly mention the interaction with the name shadowing warning, I was surprised to not see it mentioned.

I have this in the *Effect and Interactions* section:

> Top-level bindings that shadow imported names should be regarding as shadowing bindings for the purposes of `-Wname-shadowing`.

Do you think it needs to be somewhere more prominent than that?

**Comment by gergoerdi** [comment]
I believe all points brought up have been addressed by the proposal, and no new comments have shown up for a month now. @adamgundry I believe this proposal is ready for review by the Committee.

**Comment by simonpj** [comment]
* Please move all of the material in Section 4 (Interactions) into Section 2 (specification).  All three points are *specification*; they are not *consequences*.

* I would begin Section 2 by saying "When `-XImportShadowing` is enabled, the following changes take place" or something like that.

* I'd divide the spec (Section 2) with sub-headings:
  * 2.1.   References in the body of the module (the current text between 2 and 2.1)_
  * 2.2  Export lists (the first two bullets from 4)
  * 2.3  Warnings (the last bullet from 4)

* The new 2.2 (was 4) export lists.  Don't say "it feels natural".  Just say what the spec is.   Maybe "Names in export lists are resolved as specified in 2.1.   For example..."

* In the new 2.2, for "module M" item, I think you intend that the current rule continues unchanged.  The [Haskell Report says](https://www.haskell.org/onlinereport/haskell2010/haskellch5.html#x11-1000005.2) *"The form “module M” names the set of all entities that are in scope with both an unqualified name “e” and a qualified name “M.e”. "*   I think you intend the same.  Please just say that (quoting verbatim) and give examples.   (It is fine for the spec to say "There is no change to the existing spec, which is this"; that can be better than not-saying-anything.)

* "If both module B and foo are exported, then that is the same category of error as without this extension exporting module B and module C with conflicting names, and should be reported the same way."  I could not work out what this means. Please give an example.

**Comment by simonpj** [comment]
I am supportive of this proposal.

**Comment by erikd** [comment]
This proposal seems totally reasonable to me.

File: proposals/0000-import-shadowing.rst [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+   in
+     -- exports of Mod
+     (fun1, fun2)
+
+Export lists
+~~~~~~~~~~~~
+
+References in a module's export specification are resolved in the same
+scope as that used for references in the module body, as per
+:ref:`spec-body`. For example if we have something like
```
Comment (aspiwack) [inline proposals/0000-import-shadowing.rst:255]: This link doesn't render, you probably want to fix it.

File: proposals/0000-import-shadowing.rst [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+can lead to breakage just by virtue of existing code defining and
+using top-level definitions with the same name. With
+``ImportShadowing``, the existing intra-module references keep their
+meaning and there is no migration needed to accomodate the new
+``Prelude`` names.
+
+Proposed Change Specification
+-----------------------------
+
+When ``ImportShadowing`` is enabled, the following changes take place:
```
Comment (aspiwack) [inline proposals/0000-import-shadowing.rst:88]: You say so in your motivation section, but for clarity, please state clearly in this section that `ImportShadowing` is a new extension that this proposal creates.

File: proposals/0000-import-shadowing.rst [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+Consider an occurrence of a qualified name ``M.x``:
+
+(A) If the module is called ``M`` and there is a top-level binding of
+    ``x``, the occurrence is resolved to that binding
+
+(B) If the import declarations bring into scope a unique entity with
+    qualified name ``M.x``, the occurrence is resolved to that
+    entity.
+
+In both cases, Haskell 2010 regards cases (A) and (B) on equal
```
Comment (sgraf812) [inline proposals/0000-import-shadowing.rst:112]: Please link to [Haskell 2010 report, Section 5.5.2](https://www.haskell.org/onlinereport/haskell2010/haskellch5.html#x11-1090005.5.2).

**Review by sgraf812 — APPROVED** [review]
I'm in support as well.

**Review by sgraf812 — CHANGES_REQUESTED** [review]
(Apologies, I had approved the PR, but that call is not mine to make. Hence I re-requested a review by myself.)

**Review by sgraf812 — COMMENTED** [review]
Thanks. LGTM!

**Comment by gergoerdi** [comment]
I see that on `master` it is still numbered `0000`. Do I need to create a second PR that assigns it the next sequential number? Or how does that work, will magic fairies take care of that?

**Comment by JakobBruenker** [comment]
@gergoerdi I took care of it :)

**Comment by gergoerdi** [comment]
I've created https://gitlab.haskell.org/ghc/ghc/-/issues/25187 for the implementation.

**Comment by augustss** [comment]
I hadn't noticed this proposal befpre, but I'm 100% in favor.
I suggested this about 15 years ago (while implementing Mu), but sadly the idea was shot down by one single objector.

**Comment by gergoerdi** [comment]
> I suggested this about 15 years ago (while implementing Mu)

Gee I wonder where this current proposal came from 😜
